### PR TITLE
linux-v4l2: Add auto reset on timeout option

### PR DESCRIPTION
**Pull request by esden**
### Description

This pull request adds a setting to the linux-v4l2 plugin to enable automatic device resets. (Not enabled by default)

When the `select` times out (no frames available from the video capture device) for longer than the set amount of "frame times" (minimum 2, maximum 120) the plugin disable `streaming` and re-enable it again. Effectively this resets the video capture device.

### Motivation and Context

The automatic reset feature, is very useful with devices like the Elgato Cam Link 4k in bulk mode on some USB chip sets, as well as Magewell USB Capture HDMI in general. Those like to "lock up" or "freeze". 

Note: (For those that randomly find this Pull Request.) A huge stability improvement for Elgato Cam Link 4k on Linux is switching it to Isochronous mode using the Windows tool.

I left the automatic resets off by default because it might cause issues, in cases where the device is timing out a lot, but still work. (aka. dropping a lot of frames) In other cases where the devices tend to freeze, the automatic reset makes a completely unusable streaming setup, usable again with just a few dropped frames.

The timeout is calculated in the terms of frame time multiples. The reason is, that a "reasonable" timeout (not too long and not too short), is tightly coupled to the framerate of the video captured. The "frame time" multiplier can be set per device depending on how reliable they are. Experiments have shown that 5x frame times is a reasonable compromise between reaction time and false positive resets in most situations. (that is why the default is 5x frame times)

### How Has This Been Tested?
Many hours of video recording and streaming in multiple configurations.

### Types of changes

- New feature (non-breaking change which adds functionality)
- Performance enhancement (non-breaking change which improves efficiency)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [ ] I have included updates to all appropriate documentation.

### CODE REVIEW

File: plugins/linux-v4l2/v4l2-input.c [hunk 1]
```diff
@@ -796,6 +827,13 @@ static obs_properties_t *v4l2_properties(void *vptr)
 	obs_properties_add_bool(props, "buffering",
 				obs_module_text("UseBuffering"));
 
+	obs_properties_add_bool(props, "auto_reset",
+				obs_module_text("AutoresetOnTimeout"));
```
Comment (kkartaltepe) [inline plugins/linux-v4l2/v4l2-input.c:831]: Include the en-US locale strings for these.

Comment (esden) [reply plugins/linux-v4l2/v4l2-input.c:831]: Ohh I misunderstood the previous comment I guess. I though I should leave out all locale strings. :)

Comment (Fenrirthviti) [reply plugins/linux-v4l2/v4l2-input.c:831]: en-US are kept updated in-repo, every other locale is handled through CrowdIn.

**Comment by emaste** [comment]
#4004 and #4005 conflict with each other?

**Comment by kkartaltepe** [comment]
author mentioned that they touched the same lines, so yes its probable.

**Comment by esden** [comment]
@emaste as @kkartaltepe mentioned: I did mention in #4000 that these patches touch the same region of code, so to satisfy the requirement for granular pull requests, I had the following options:
* the two pull requests will conflict
* the two patches will depend on each other
* or keep it all as one pull request 

I hope that having the patches be separate will speed up them being merged. 🤞

@emaste if you want to try both patches together you can use this branch here: https://github.com/esden/obs-studio/tree/v4l-robustness-improvements

**Comment by esden** [comment]
As #4004 is now merged. I have rebased this patch so that it does not conflict any more.

**Comment by jp9000** [comment]
Why specifically do these devices "like to lock up" as you say?

**Comment by jp9000** [comment]
Sorry, I read the description, but I suppose I'm not an expert on Linux matters, so I'm trying to understand the situation. Was this something we were doing wrong from the beginning, or is this something broken on their end that we're trying to hack in a fix for?

**Comment by esden** [comment]
@jp9000 I wish I had a good answer for you. The goal of my investigation was to find out the root cause. Unfortunately I was unable to figure out what the exact bug is. All I can tell you is: "this patch helps a lot to make things much more reliable and way less annoying" :D 

Depending on the device they might stop streaming frames and "freeze" for different reasons.

Magewell USB Capture HDMI Gen 2: Freezes randomly. Usually at the start of obs capture. Then runs stable for an hour or two and then freezes again. The resets introduced in this patch prevent the device from freezing permanently, so it ends up being a short glitch instead of a frozen frame.

Elgato Cam Link 4K: Ships from factory using bulk endpoint USB. This works more or less reliably on some computers (my old streaming setup using a Lenovo Laptop) but with other setups (my new AMD Ryzen desktop) freezes almost immediately after only 1-2 minutes of OBS capturing the stream. After switching the device to Isochronous mode using the windows software it is much more reliable. Still if I forget to switch on my camera, that is connected to the Elgato, and start OBS, in about 50% of the cases the `select` will time out and freeze permanently until I restart OBS. Even switching on the camera does not help to recover the device and I have to restart OBS. This patch resolves this issue and resets the device until I switch on the camera.

Even though despite my best efforts I could not figure out what the exact deep underlying issue is. It could be a freeze and a bug on the device side, it could be a kernel driver issue, it could be a USB stack issue. Especially because triggering a freeze can be induced in many different ways and with both of the USB capture devices that I have. This patch even with the Elgato device switched to isochronous (that is much more robust than bulk), takes care of really annoying video freezes in my streaming setup. Having video freeze permanently and require an OBS restart in the middle of a 2 hour stream is annoying at best. And based on reports from other OBS Linux users, they have similar issues to mine that they work around and are sometimes able to get it stable enough but never perfect. So I hope this patch will be another tool in the toolbox to stabilize the use of OBS on linux.

For additional information. I did also test with guvcview, capturing all the devices that I use at the same time, to provide same USB stack load. guvcview never encounters freezes the same way OBS does. I did read the sourcecode of guvcview and they don't implement any reset mechanisms. So there still might be some hard to debug multiprocess timing issue or other interaction between audio USB devices and video USB devices. But debugging this is way beyond my expertise and time that I can invest into debugging it.

So at the end, again I want to say, this patch might be at the end a workaround for some obscure bug in OBS (considering that guvcview does not have an issue like this) but it is way better than having randomly freezing video streams. :)

I hope this helps.

**Pull request merged**